src/runtime/auto-prompt-manager.ts

```diff
@@ -579,6 +607,64 @@ export class AutoPromptManager {
     return actionToUse;
   }
 
+  private async getPotentialAction_({
+    autoPromptType,
+    article,
+    frequencyCapConfig,
+  }: {
+    autoPromptType: AutoPromptType;
+    article: Article;
+    frequencyCapConfig: FrequencyCapConfig | undefined;
+  }): Promise<Intervention | void> {
+    let actions = article.audienceActions?.actions;
+    if (!actions) {
+      return;
+    }
+
+    const isSurveyEligible = await this.isSurveyEligible_(actions);
+    actions = actions.filter((action) =>
+      this.checkActionEligibility_(
+        action.type,
+        // Monetization check does not apply for new frequency capping flow
+        /** canDisplayMonetizationPrompt */ true,
+        isSurveyEligible
+      )
+    );
+
+    // FrequencyCapConfig may be undefined ONLY for subscription openacess
+    // content. If so, display first valid action.
+    if (
+      this.isSubscription_(autoPromptType) ||
+      !this.isValidFrequencyCap_(frequencyCapConfig)
+    ) {
+      return actions[0];
+    }
+
+    const globalFrequencyCapDuration =
+      frequencyCapConfig?.globalFrequencyCap?.frequencyCapDuration;
+    if (this.isValidFrequencyCapDuration_(globalFrequencyCapDuration)) {
+      const globalImpressions = await this.getAllImpressions_();
+      if (
+        this.isFrequencyCapped_(globalFrequencyCapDuration!, globalImpressions)
+      ) {
+        return;
+      }
+    }
+
+    for (const action of actions) {
```

Comment:
for this implementation, it is possible for impressions for each action to be fetched twice: once for the global frequency cap, and then again for each individual prompt frequency cap. 

An alternative is to save the results of the impressions fetch for the global cap into a map and reuse that for the prompt frequency cap. What are your thoughts?

Reply:
> for each action to be fetched twice

is that referring to reading data from local stroage? As far as I know, cost of accessing local storage is significantly low unless the data is large, so I think it's negligible. I don't have much concern on performance and would prefer having simpler code.

Reply:
Agree, the implementation is a bit cleaner, the tests will need to check for the additional local storage fetches but I think it is still preferable.